src/internal/pachd/builder.go

```diff
@@ -548,6 +556,64 @@ func (b *builder) startDebugWorker(ctx context.Context) error {
 	return nil
 }
 
+func (b *builder) generatePJSWorkerAuthToken(ctx context.Context) error {
+	return generatePJSWorkerAuthToken(ctx, b.env.AuthServer(), b.env.GetKubeClient(), b.env.Config())
+}
+
+// TODO: move this function to the auth package(s) it can potentially be called again when we turn on auth.
+func generatePJSWorkerAuthToken(ctx context.Context, authServer auth.APIServer, kubeClient kubernetes.Interface,
+	config *pachconfig.Configuration) error {
+	secretName := "pachyderm-pjs-worker-authentication"
+	secretKey := "pjsWorkerAuthToken"
+	// First read the secret in case it exists.
+	existingSecret, err := kubeClient.CoreV1().Secrets(config.Namespace).Get(context.Background(), secretName, metav1.GetOptions{})
+	if err == nil {
+		// Secret exists, read the value from the specified key
+		if encodedValue, ok := existingSecret.Data[secretKey]; ok {
+			// Decode the value since it is stored as base64
+			token, err := base64.StdEncoding.DecodeString(string(encodedValue))
+			if err != nil {
+				return errors.Wrap(err, "decoding from base64")
+			}
+			if err := os.Setenv("PJS_WORKER_AUTH_TOKEN", string(token)); err != nil {
+				return errors.Wrap(err, "set the token as an environment variable")
+			}
+			config.PJSWorkerAuthToken = string(token)
+			return nil
+		}
+		// TODO: delete and recreate?
+		return errors.New(secretName + " does not contain expect secret key")
+	} else if !k8serrors.IsNotFound(err) {
+		return errors.Wrap(err, "getting pjs worker auth secret")
+	}
+	// create the secret if it doesn't exist.
+	ctx = internalauth.AsInternalUser(ctx, authdb.InternalUser)
+	tokenResp, err := authServer.GetRobotToken(ctx, &auth.GetRobotTokenRequest{Robot: "pjs-worker"})
+	if err != nil {
+		return errors.Wrap(err, "getting robot token")
+	}
+	//TODO: TTL?
+	encodedSecret := base64.StdEncoding.EncodeToString([]byte(tokenResp.Token))
+	secret := &v1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: secretName,
+		},
+		Data: map[string][]byte{
+			secretKey: []byte(encodedSecret),
+		},
+		Type: v1.SecretTypeOpaque,
+	}
+	_, err = kubeClient.CoreV1().Secrets(config.Namespace).Create(context.Background(), secret, metav1.CreateOptions{})
+	if err != nil {
+		return errors.Wrap(err, "creating kubernetes secret")
+	}
+	if err := os.Setenv("PJS_WORKER_AUTH_TOKEN", tokenResp.Token); err != nil {
```

Comment:
Ditto about setting the token here. No real need to set it in the process’s environment, since the process already knows it, and `pachd` doesn’t spawn new processes, right?